File: typical90/day01.py
n,l=map(int,input().split())
k=int(input())
a=[*map(int,input().split())]
def is_ok(x,k):
    b=0
    m=0
    for i in a:
        if i-b>=x and l-i>=x:
            b=i
            m+=1
    return m>=k
ok=0
ng=l+1
while abs(ok-ng)>1:
    mid=(ok+ng)//2
    if is_ok(mid,k):
        ok=mid
    else:
        ng=mid
print(ok)

File: rust/src/bin/day01.rs
use proconio::input;

fn main() {
    input! {
        n: isize,
        l: isize,
        k: isize,
        a: [isize; n]
    }

    let mut ok: isize = 1;
    let mut ng: isize = l + 1;
    while (ng - ok).abs() > 1 {
        let mid = (ng + ok) / 2;
        let mut pre = 0;
        let mut cnt = 0;
        for i in &a {
            if i - pre >= mid && l - i >= mid {
                pre = *i;
                cnt += 1;
            }
        }
        if cnt >= k {
            ok = mid;
        } else {
            ng = mid;
        }
    }
    println!("{}", ok);
}
